drivers/hmis/app/models/hmis/hud/custom_service_category.rb

```diff
@@ -14,7 +14,7 @@ class Hmis::Hud::CustomServiceCategory < Hmis::Hud::Base
   belongs_to :user, **hmis_relation(:UserID, 'User'), optional: true
   has_many :service_types, class_name: 'Hmis::Hud::CustomServiceType'
   has_many :form_instances, class_name: 'Hmis::Form::Instance'
-  has_many :definitions, through: :form_instances
+  has_many :definitions, through: :form_instances, source: :definition # todo @martha - `definitions` is probably clearer, but `definition` would be the one that doesn't change behavior
```

Comment:
Ah interesting, I'm curious which test caught this?

Reply:
[drivers/hmis/spec/requests/hmis/service_category_spec.rb](https://github.com/greenriver/hmis-warehouse/blob/5eed1631d1ae87951eb01bb1b4d4e7d064142b27/drivers/hmis/spec/requests/hmis/service_category_spec.rb#L73-L80)

```
Hmis::GraphqlController when the user has access should successfully create a service category
     Failure/Error: expect(response.status).to eq(200), result.inspect
       {"errors"=>[{"message"=>"Ambiguous source reflection for through association. Please specify a :source directive on your declaration like:\n\n  class Hmis::Hud::CustomServiceCategory < ActiveRecord::Base\n    has_many :definitions, {:through=>:form_instances, :source=>\"definition\"}\n  end", "backtrace"=>["/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/reflection.rb:930:in `source_reflection_name'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/reflection.rb:822:in `source_reflection'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/reflection.rb:963:in `check_validity!'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/associations/association.rb:41:in `initialize'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/associations/has_many_through_association.rb:10:in `initialize'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/associations.rb:304:in `new'", "/bundle/ruby/3.1.0/gems/activerecord-7.0.8.4/lib/active_record/associations.rb:304:in `association'", "/app/drivers/hmis/app/graphql/sources/active_record_association.rb:24:in `block (2 levels) in fetch'", "/app/drivers/hmis/app/graphql/sources/active_record_association.rb:23:in `each'", "/app/drivers/hmis/app/graphql/sources/active_record_association.rb:23:in `block in fetch'", "/app/drivers/hmis/app/graphql/sources/active_record_association.rb:22:in `each'", "/app/drivers/hmis/app/graphql/sources/active_record_association.rb:22:in `fetch'", "/bundle/ruby/3.1.0/gems/graphql-2.3.1/lib/graphql/dataloader/source.rb:136:in `run_pending_keys'", "/bundle/ruby/3.1.0/gems/graphql-2.3.1/lib/graphql/dataloader.rb:271:in `each'", "/bundle/ruby/3.1.0/gems/graphql-2.3.1/lib/graphql/dataloader.rb:271:in `block in spawn_source_fiber'", "/bundle/ruby/3.1.0/gems/graphql-2.3.1/lib/graphql/dataloader.rb:233:in `block in spawn_fiber'"]}], "data"=>{}}
     # ./drivers/hmis/spec/requests/hmis/service_category_spec.rb:75:in `block (3 levels) in <top (required)>'
     # /bundle/ruby/3.1.0/gems/webmock-3.23.0/lib/webmock/rspec.rb:39:in `block (2 levels) in <main>'
```

[drivers/hmis/spec/requests/hmis/project_services_spec.rb](https://github.com/greenriver/hmis-warehouse/blob/5eed1631d1ae87951eb01bb1b4d4e7d064142b27/drivers/hmis/spec/requests/hmis/project_services_spec.rb#L112-L127) - both tests in here also fail with a similar error.

Reply:
Oh I see, thanks. Its just from loading the model. I looked back at [when this relation was added](https://github.com/greenriver/hmis-warehouse/commit/06ec5c2b275fe415474a9c197ed35ccbe07ab701#diff-0694f7e51774a198ad58178bd4c2d8e780a4bdf3d7a25f7a9bc18ed810c36482). Its used to resolve the list of forms that are used for a given service type/category, here: https://github.com/greenriver/hmis-warehouse/blob/652e046dbd37ed7f54602c7db023ccd988ce218b/drivers/hmis/app/graphql/types/hmis_schema/service_type.rb#L42-L48


I think we probably want to add a similar `select(&:published?)` to that at the end, so we're not repeating the form definition for each version that it has (And not including draft/retired forms). I'm not super concerned about what is being resolved there since it's only informational for the config tool page.

Reply:
Oh, makes sense, thanks for digging into it!